### ISSUE TITLE
Remove references to the internals of LedgerTxnEntry

### ISSUE THREAD

**Issue by sisuresh** [stellar/stellar-core#2921]
In some operations like `MergeOp` and `SetOptionsOp`, a reference to the internals of `LedgerTxnEntry` is held. This is a dangerous pattern to follow since the deactivation of the  `LedgerTxnEntry` could lead to undefined behavior.

Example - https://github.com/stellar/stellar-core/blob/85da7712fe9654ecf5309b95f109552f9fff48dd/src/transactions/SetOptionsOpFrame.cpp#L110

We should audit the code and remove these references. One solution that @jonjove suggested was to directly use a lambda that returns a reference, so the reference is never stored.

**Comment by MonsieurNicolas** [comment]
This sounds reasonable to avoid UB, but that pattern "only" gets us to "internal error".

```
LedgerTxnEntry x = txn.load(...);
auto& rx = x.current();
// use rx
cleverCode1(rx);
foo(txn); // maybe invalidates x
// use rx again
cleverCode2(rx); // UB if x got invalidated
```

is now

```
LedgerTxnEntry x = txn.load(...);
// use x
cleverCode1(x.current());
foo(txn); // maybe invalidates x
// use x again
cleverCode2(x.current()); // throws if x got invalidated
```

I think that the problem here is that we "don't know" what `foo` is going to do (it may or may not invalidate).

So *maybe* the fix here (short of trying to do something clever with scope), would be to add a new method to `LedgerTxn` that "deactivates all entries" performs the equivalent of (this is basicaly `addChild` without the `mChild` management):
```
     AbstractLedgerTxn& LedgerTxn::Impl::deactivateAll()
     {
          {
               LedgerTxn txn(*this);
          }
          return *this;
     }
```

actually, we may want to make this mandatory in code outside of `LegderTxn*`:
the only way to get a reference to a `AbstractLedgerTxn&` from a `LedgerTxn` instance is to either call `deactivateAll` or call some method `getRawReference` (that is unsafe).

**Comment by jonjove** [comment]
I agree that this "only" goes from UB to internal error, but that is still a huge improvement. The motivation for doing this is that references like `auto& le = ltxe.current()` intrinsically bypass the safety mechanisms of `LedgerTxn`. So we should continue using this lambda pattern for now.

But I'm very receptive to the idea of building an interface that makes it harder (ideally impossible) to have this problem. Right now, I don't know what the best approach would be. It would be best if we could guarantee more at compile time.

**Comment by jonjove** [comment]
A bit of searching on the internet just turned up https://clang.llvm.org/docs/AttributeReference.html#consumed-annotation-checking which seems like it could be useful in this context. It's a clang extension so I think we would need to put it behind conditionally compiled macros. The benefits of something like this could be significant, but I'm uncertain of the burden to set it up (there is a lot of code using `LedgerTxn` and related classes, will we have to annotate everything everywhere?).

**Comment by MonsieurNicolas** [comment]
The problem is actually worst.

@sisuresh run into this while following the lambda pattern, but I am describing here the general unsafe pattern in our code base.

Consider this:
```
    removeEntryWithPossibleSponsorship(
        ltx, header, sourceAccountEntry.current(), sourceAccountEntry);
```
that looks pretty safe... except that if that function creates a child `LedgerTxn`, then `sourceAccountEntry.current()` can be arbitrary invalidated because `sourceAccountEntry.current()` doesn't actually force to hold a reference to the actual ledger entry. 

So basically the reference returned by `sourceAccountEntry.current()` is held alive by some other data structure (`mActive` or `mEntry` in `ltx`).

The danger here is that if `sourceAccountEntry` in this example was loaded "without record", then the underlying object will get deleted, which will cause the reference to point to garbage (use after free type of error).

I *think* this is the only way the reference can be invalidated - but maybe not (like when dealing with root)?

**Comment by MonsieurNicolas** [comment]
I think that we could make this safer/easier to see the bug if there was a way to pass around `SealedLedgerTxn` that cannot create a child? Many helper functions load a bunch of stuff without creating a child txn so it's actually safe to follow the pattern above; on the other hand, they are exposed to hitting a "already active" type of bug.
